File: apps/api/tests/recorded.py
"""Transports that replay recorded upstream responses. Any unrecorded request fails."""

import hashlib
import json
from pathlib import Path

import httpx

FIXTURES = Path(__file__).parent / "fixtures"


def _rxnorm(request: httpx.Request) -> httpx.Response:
    path = request.url.path
    if path.endswith("/rxcui.json"):
        name = request.url.params["name"].lower().replace(" ", "_")
        fixture = FIXTURES / "rxnorm" / f"rxcui_{name}.json"
    elif path.endswith("/related.json"):
        fixture = FIXTURES / "rxnorm" / f"related_{path.split('/')[-2]}.json"
    else:
        raise AssertionError(f"Unexpected request: {request.url}")
    return httpx.Response(200, json=json.loads(fixture.read_text()))


def _openfda(request: httpx.Request) -> httpx.Response:
    # Names match scripts/record_openfda.py.
    params = request.url.params
    key = f"{params['search']}|{params['skip']}|{params['limit']}"
    fixture = FIXTURES / "openfda" / (hashlib.sha1(key.encode()).hexdigest()[:12] + ".json")
    if not fixture.exists():
        raise AssertionError(f"Unrecorded request: {params}")
    envelope = json.loads(fixture.read_text())
    return httpx.Response(envelope["status"], json=envelope["body"])


def rxnorm_http() -> httpx.Client:
    return httpx.Client(transport=httpx.MockTransport(_rxnorm), base_url="https://rxnav.test")


def openfda_http() -> httpx.Client:
    return httpx.Client(transport=httpx.MockTransport(_openfda), base_url="https://fda.test")

File: apps/worker/test/recorded.ts
import type { Fetch } from '../src/http.ts'

// Replays recorded upstream responses. Any unrecorded request fails the test.

const FIXTURES = import.meta.glob<unknown>('../../../fixtures/{rxnorm,openfda}/*.json', {
  eager: true,
  import: 'default',
})

function fixture(name: string): unknown {
  return FIXTURES[`../../../fixtures/${name}`]
}

export function json(status: number, body: unknown): Response {
  return new Response(JSON.stringify(body), {
    status,
    headers: { 'Content-Type': 'application/json' },
  })
}

/** A fetch answered by `handler`, which sees the parsed URL. */
export function mockFetch(handler: (url: URL) => Response | Promise<Response>): Fetch {
  return async (input) => handler(new URL(input))
}

export function rxnormFetch(): Fetch {
  return mockFetch((url) => {
    const path = url.pathname
    let name: string
    if (path.endsWith('/rxcui.json')) {
      const query = (url.searchParams.get('name') ?? '').toLowerCase().replaceAll(' ', '_')
      name = `rxnorm/rxcui_${query}.json`
    } else if (path.endsWith('/related.json')) {
      name = `rxnorm/related_${path.split('/').at(-2)}.json`
    } else if (path.endsWith('/displaynames.json')) {
      name = 'rxnorm/displaynames.json'
    } else {
      throw new Error(`Unexpected request: ${url}`)
    }
    const body = fixture(name)
    if (body === undefined) throw new Error(`Unrecorded request: ${url}`)
    return json(200, body)
  })
}

async function sha1(text: string): Promise<string> {
  const digest = await crypto.subtle.digest('SHA-1', new TextEncoder().encode(text))
  return [...new Uint8Array(digest)].map((b) => b.toString(16).padStart(2, '0')).join('')
}

type Envelope = { status: number; body: unknown }

export function openfdaFetch(): Fetch {
  // Names match apps/api/scripts/record_openfda.py.
  return mockFetch(async (url) => {
    const p = url.searchParams
    const key = `${p.get('search')}|${p.get('skip')}|${p.get('limit')}`
    const envelope = fixture(`openfda/${(await sha1(key)).slice(0, 12)}.json`) as
      | Envelope
      | undefined
    if (envelope === undefined) throw new Error(`Unrecorded request: ${url}`)
    return json(envelope.status, envelope.body)
  })
}

/** A fetch that fails the test if called at all. */
export function unreachable(): Fetch {
  return mockFetch((url) => {
    throw new Error(`Unexpected upstream request: ${url}`)
  })
}
